src/mame/brother/lw30.cpp

```diff
@@ -0,0 +1,1105 @@
+		return ~out;
+	}
+
+	// Floppy
+	TIMER_DEVICE_CALLBACK_MEMBER(floppy_timer_callback) {
+		auto floppy_device = floppy->get_device();
+		if(floppy_device->ready_r() != false)
+			return;
+
+		floppy_latch <<= 1;
+
+		attotime now = machine().time();
+		attotime when = now - attotime::from_usec(4);
+		attotime reversal = floppy_device->get_next_transition(when);
+		if(reversal > when && reversal <= now)
+			floppy_latch |= 1;
+
+		floppy_shifter++;
+		if((floppy_read_until_zerobit && (floppy_latch & 1) == 0) || (!floppy_read_until_zerobit && floppy_shifter == 8)) {
+			//LOGFLOPPY("%s: floppy_timer_callback: floppy_read_until_zerobit=%d latch=%02X\n", machine().describe_context(), floppy_read_until_zerobit, floppy_latch);
+			floppy_control |= 0x80; // floppy_data_available = true;
+			floppy_data = floppy_latch;
+			floppy_latch = floppy_shifter = 0;
+			floppy_read_until_zerobit = false;
+		}
+		//LOGFLOPPY("%s: floppy_timer_callback: IO_80=%02X, shifter=%u offset=%u\n", machine().describe_context(), io_80, floppy_shifter, floppy_track_offset % cache.size());
+		//LOGFLOPPY("%s: read_io_80 track=%d,offset=%4x => %02x\n", callstack(), floppy_steps / 4, floppy_track_offset, ret);
+	}
+
+	uint8_t floppy_data_r() { // 80
+		floppy_control &= ~0x80; // floppy_data_available = false;
+		LOGFLOPPY("%s: read %02X from IO 80\n", machine().describe_context(), floppy_data);
+		return floppy_data;
+	}
+	void floppy_data_w(uint8_t data) {
+		LOGFLOPPY("%s: write %02X to IO 80\n", machine().describe_context(), data);
+		floppy_data = data;
+	}
+
+	uint8_t io_88_r() {
+		// bit 0: set in start_write; cleared in end_write
+		// bit 1: pulsed after 3*0xFF sync (read next floppydata until zero-bit)
+		// bit 2: cleared in stepper routines, rst28_06
+		// bit 3: set in start_write; cleared in end_write
+		// bit 5: cleared in rst28_06; motor-on?
+		LOGFLOPPY("%s: read %02X from IO 88\n", machine().describe_context(), io_88);
+		return io_88;
+	}
+	void io_88_w(uint8_t data) {
+		LOGFLOPPY("%s: write %02X to IO 88\n", machine().describe_context(), data);
+		io_88 = data;
+		floppy->get_device()->mon_w((io_88 & (1 << 5)) == 0);
+	}
+
+	uint8_t floppy_status_r() { // 90
+		// bit 7 set; data ready from floppy
+		// bit 6 clear; unknown meaning
+		// bit 5 clear; unknown meaning
+		// bit 4 clear; unknown meaning
+		// bit 3-0: stepper motor
+		LOGFLOPPY("%s: read %02X from IO 90\n", machine().describe_context(), floppy_control);
+		return floppy_control;
+	}
+	void floppy_stepper_w(uint8_t data) {
+		LOGFLOPPY("%s: write %02X to IO 90\n", machine().describe_context(), data);
+		// write directly to 4-wire bipolar stepper motor (see stepper_table)
+		// a rotation to the left means decrease quarter-track
+		// a rotation to the right means increase quarter-track
+		auto rol4 = [](uint8_t d) { return ((d << 1) & 0b1111) | ((d >> 3) & 0b0001); };
+		auto ror4 = [](uint8_t d) { return ((d >> 1) & 0b0111) | ((d << 3) & 0b1000); };
+		auto old_track = floppy_steps / 4;
```

Comment:
Can you make some of these locals `const`?  The lambdas at least aren’t reassigned and have no mutable state.